lighthouse-core/fraggle-rock/gather/driver.js

```diff
@@ -0,0 +1,64 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const ProtocolSession = require('./session.js');
+const RuntimeController = require('./runtime-controller.js');
+
+class Driver {
+  /**
+   * @param {import('puppeteer').Page} page
+   */
+  constructor(page) {
+    this._page = page;
+    /** @type {LH.Gatherer.FRProtocolSession|undefined} */
+    this._session = undefined;
+    /** @type {RuntimeController|undefined} */
+    this._runtimeController = undefined;
+
+    // Proxy the protocol session methods for legacy compat.
+    const getSession = () => {
+      if (!this._session) throw new Error(`Driver not connected to page`);
+      return this._session;
+    };
+
+    /** @type {LH.Gatherer.FRTransitionalDriver['hasNextProtocolTimeout']} */
+    this.hasNextProtocolTimeout = (...args) => getSession().hasNextProtocolTimeout(...args);
+    /** @type {LH.Gatherer.FRTransitionalDriver['getNextProtocolTimeout']} */
+    this.getNextProtocolTimeout = (...args) => getSession().getNextProtocolTimeout(...args);
+    /** @type {LH.Gatherer.FRTransitionalDriver['setNextProtocolTimeout']} */
+    this.setNextProtocolTimeout = (...args) => getSession().setNextProtocolTimeout(...args);
+    /** @type {LH.Gatherer.FRTransitionalDriver['on']} */
+    this.on = (...args) => getSession().on(...args);
+    /** @type {LH.Gatherer.FRTransitionalDriver['off']} */
+    this.off = (...args) => getSession().off(...args);
+    /** @type {LH.Gatherer.FRTransitionalDriver['sendCommand']} */
+    this.sendCommand = (...args) => getSession().sendCommand(...args);
+
+    /** @type {LH.Gatherer.FRTransitionalDriver} */
+    const typecheck = this; // eslint-disable-line no-unused-vars
+  }
+
+  /** @return {Promise<void>} */
+  async connect() {
+    if (this._session) return;
+    const session = await this._page.target().createCDPSession();
```

Comment:
Speaking of rethinking some of the foundations: it was a super ancient architecture decision that's long gone that made it so `Driver` was the one initiating the protocol connection in `Driver.connect()` (vs attaching in `GatherRunner` or `index.js` or whoever was passing in the connection). And even though we still go through the motions, we broke it with DevTools connections (`Raw` connections are already attached so `connect()` is just a lie).

Would it simplify things in Driver to throw that away definitively by constructing with a `Session` instead? If a driver is only usable when it has a session, there may be no reason to make it possible to end up with one without a session, and it would let us drop all of the "Driver not connected to page" checks.

Reply:
It was actually fortuitous :) we will need on demand session creation like this, so we it's good that we have it this way based on a `page`